File: Heuristic Search Planner (sgware)/src/edu/uno/ai/planning/hsp/MaxHeuristic.java
package edu.uno.ai.planning.hsp;

import edu.uno.ai.planning.logic.Conjunction;
import edu.uno.ai.planning.logic.Expression;
import edu.uno.ai.planning.logic.Literal;
import edu.uno.ai.planning.ss.StateSpaceProblem;

public class MaxHeuristic extends HSPHeuristic {

	public MaxHeuristic(StateSpaceProblem problem) {
		super(problem);
	}

	@Override
	protected int cost(Expression expression) {
		if(expression instanceof Literal)
			return cost((Literal) expression);
		else if(expression instanceof Conjunction) {
			int max = 0;
			for(Expression argument : ((Conjunction) expression).arguments) {
				int argCost = cost(argument);
				if(argCost == INFINITY)
					return INFINITY;
				else
					max = Math.max(max, argCost);
			}
			return max;
		}
		else
			throw new UnsupportedOperationException(expression.getClass() + " not supported.");
	}
}

File: UnitySokoban/Assets/Scripts/Planning/HeuristicSearchPlannerSGW/MaxHeuristic.cs
using StateSpaceSearchProject;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Planning.Logic;

namespace HeuristicSearchPlannerSGW
{
    public class MaxHeuristic : HSPHeuristic
    {
        public MaxHeuristic(StateSpaceProblem problem)
            : base(problem)
        {
        }

        protected override int cost(Expression expression)
        {
            if (expression is Literal)
                return cost((Literal)expression);
            else if (expression is Conjunction)
            {
                int max = 0;
                foreach (Expression argument in ((Conjunction)expression).arguments)
                {
                    int argCost = cost(argument);
                    if (argCost == INFINITY)
                        return INFINITY;
                    else
                        max = Math.Max(max, argCost);
                }
                return max;
            }
            else
                throw new InvalidOperationException(expression.GetType() + " not supported.");
        }
    }

}
